### ISSUE TITLE
Re-land "add IntoFuture trait and support for await"

### ISSUE THREAD

**Issue by tmandry** [rust-lang/rust#67982]
#65244 was reverted in https://github.com/rust-lang/rust/pull/67768 for performance reasons (see https://github.com/rust-lang/rust/issues/67706). This issue tracks re-landing the support in a more performant way.

Tracking issue for `IntoFuture` trait:  https://github.com/rust-lang/rust/issues/67644

cc @seanmonstar @wesleywiser

**Comment by seanmonstar** [comment]
Sorry for my lack of rustc internal knowledge, is there any hint or guess at *why* it was slower?

**Comment by tmandry** [comment]
I don't think anyone knows for sure at this point. https://github.com/rust-lang/rust/issues/67706 mentions `normalize_ty_after_erasing_regions` as the query that regressed.

cc @rust-lang/wg-traits

**Comment by tmandry** [comment]
In the @rust-lang/wg-async-await meeting we agreed that this was a nice to have, but would be good to get it over the finish line. Tagging this as OnDeck, because we'd like to look into it while people still remember the implementation details.

**Comment by jonas-schievink** [comment]
The same query regressed by 10% in https://github.com/rust-lang/rust/pull/68524, also without being invoked more often.

**Comment by bstrie** [comment]
#68606 alleges to unblock this issue, would anyone like to benchmark an un-reversion to see whether or not that's true?

**Comment by jonas-schievink** [comment]
I made further perf improvements in https://github.com/rust-lang/rust/pull/68672, so we can also wait until that lands.

**Comment by Xanewok** [comment]
The linked PR has been merged and it looks like the async/await benchmark received quite a nice boost.

Is it blocked on https://github.com/rust-lang/rust/issues/68864 to make progress?

**Comment by nikomatsakis** [comment]
I don't think it would be blocked on #68864, but maybe I'm missing something. Seems like we should re-open the PR and then maybe do some perf comparisons.

**Comment by jonas-schievink** [comment]
It has been reopened as https://github.com/rust-lang/rust/pull/68811. According to perf there's still a compile time regression of up to ~6% present.

**Comment by tmandry** [comment]
@rfcbot assign @yoshuawuyts 

Who was interested in getting at least the (unstable) trait merged. It'd be nice to do another perf run with the async/await integration, also.

**Comment by tmandry** [comment]
@rustbot release-assignment

The trait has been merged in #72459, so this issue remains to track actual integration with `.await`.

**Comment by WaffleLapkin** [comment]
What is the current status of this issue?

**Comment by Nemo157** [comment]
https://github.com/rust-lang/rust/pull/68811#issuecomment-582172770 seems to be the latest "what needs to be done next".

**Comment by danieleades** [comment]
this seems to have stalled. that's a real shame, because I have a few crates where this would massively help with ergonomics

**Comment by Type1J** [comment]
@danieleades Same here.

**Comment by wesleywiser** [comment]
There's work ongoing in #90737 to resolve this. Thanks @eholk  and @yoshuawuyts! 🎉

**Comment by camelid** [comment]
#90737 has been merged now. Should this issue be closed?

**Comment by eholk** [comment]
@camelid - Yes, the issue can be closed now. #67644 can probably be closed as well.

**Comment by camelid** [comment]
> #67644 can probably be closed as well.

AFAICT the feature hasn't been stabilized yet, so I think the tracking issue should stay open.

**Comment by eholk** [comment]
Sounds good to me!

**Comment by Type1J** [comment]
In Microsoft's `windows` crate, there are quite a few structs that implement a `poll()` function and the `IntoFuture` trait, but it only enabled `IntoFuture` on nightly. This is more their problem than Rust's, but if `IntoFuture` was in stable, it would most likely be fixed. Right now, I believe that they're in a "wait and see" mode.

**Comment by yoshuawuyts** [comment]
@Type1J Do you have a link to the `IntoFuture` implementations in the `windows` crate? I found 4 `Future` trait implementations in the entire source, but no references to `IntoFuture`? Am I missing something?

I've recently started helping out on the `windows-rs` project, so if there is a dependency on `IntoFuture` from `windows-rs` I would love to know, as that would provide a convenient opportunity outside of the HTTP client space to validate how `IntoFuture` performs.

**Comment by Type1J** [comment]
My mistake. It seems to be a blanket implementation.

https://microsoft.github.io/windows-docs-rs/doc/windows/Foundation/struct.IAsyncOperation.html

**Comment by khuey** [comment]
This caused significant regressions (i.e. 25% or more) in stack frame size in debug builds of my application. Was that expected?

**Comment by khuey** [comment]
I could extract MIR or something but the code in question is not open source.

**Comment by eholk** [comment]
It's weird that we're only seeing these regressions on debug builds. I wouldn't expect the code around this to change between debug and release.

**Comment by khuey** [comment]
Presumably the optimizer is doing its job on release builds.
